### ISSUE TITLE
Use mission control information in QueryRoutes

### ISSUE THREAD

**Issue by C-Otto** [lightningnetwork/lnd#3209]
According to the discussion in https://github.com/lightningnetwork/lnd/issues/3206, lnd does not make use of mission control information about failing routes/edges/nodes when returning a route for a QueryRoutes request. I'd like lnd to return a different route after a failed payment when using QueryRoutes.

**Comment by Roasbeef** [comment]
It _does_ take them into account (same code pathway for path finding), failures will also be incorporated.

**Comment by joostjager** [comment]
`QueryRoutes` does not take into account mission control information.

https://github.com/lightningnetwork/lnd/blob/d6d87e12feeb197f599a329c079ecef17a4f2b4a/lnrpc/routerrpc/router_backend.go#L140

There it passes in a probability source that just returns binary probabilities based on caller supplied ignore lists.

**Comment by joostjager** [comment]
@C-Otto There are currently basically two models of using `lnd` for payments:

* `SendPayment`: `lnd` handles everything. It launches multiple payment attempts if needed and keeps track of node reputations in mission control.

* `QueryRoutes` + `routerrpc.SendToRoute`: these are lower level calls that require the caller to have its own reputation management system (mission control). Callers can use the detailed error returned from `SendToRoute` to update their own reputation data. Then use the updated node reputations to build new ignore lists to pass into `QueryRoutes`. Basically replicating the logic that is behind `SendPayment` in `lnd`.

   A useful follow up could be to allow probabilities to be passed into `QueryRoutes` instead of the ignore lists.

The only point where both models touch is `SendToRoute` passing every outcome into `lnd`'s mission control. The idea is that if a user uses both models, future `SendPayment` calls can benefit from what was learnt through `SendToRoute`.

Your request in this issue is mix both models. How exactly would you be using this if it were possible?

**Comment by C-Otto** [comment]
I don't know a lot about lnd internals, and I don't want to interfere with plans or mix code that shouldn't be mixed. But I can explain my perspective.

I'd like to rebalance a specific channel. For that I need to create my own route, i.e. extend some route by one additional hop back to my own node. To find this route I need to use QueryRoutes.

However, I'm not interested in adding any mission control logic in my code. My very naive assumption is that lnd is able to figure out which nodes/edges/... are broken, and ignore them based on some clever heuristics. As such, I expect routes returned by QueryRoutes to be usable (at least when trying multiple times).

In other words, I don't see the advantage of QueryRoutes. I could just grab the whole graph and find a path myself. I'm using lnd RPC calls to take advantage of lnd, and not getting fancy mission control information is just sad.

**Comment by joostjager** [comment]
Thanks for clarifying the use case. That provides good context and I see why you'd want this change.

The advantage of `QueryRoutes` in its current form is that it allows you to implement your own mission control without having to implement your own path finding.

Extending `QueryRoutes` with an option to use mission control probabilities is not a big change. A question may be what to do when the caller also provides ignore lists (or in the future perhaps probability list). How to merge those two sources of probabilities? Of course they could also be mutual exclusive.

I am wondering though whether it isn't better to just implement route-to self in `lnd`. If that would work, would you still have a need for a different `QueryRoutes`?

**Comment by C-Otto** [comment]
I don't know anything about the first usecase (clients with their own mission control). Maybe it helps clarifying these needs. Adding mission control (with an optional switch?) sounds perfect to me. Maybe it helps to have a dedicated endpoint, `QueryBestRoute` or similar.

Ignoring channels due to high fees could be a use case for my rebalance script, though (in case lnd does not consider fees when finding routes).

Route-to-self in lnd would be even better. For that to be helpful in my use case I'd need to specify a channel (and direction) that should be used as part of the resulting transaction.

**Comment by joostjager** [comment]
The main reason that we have `QueryRoutes` with its parameters is that it allows users to implement their own k-shortest if they want to. Anyone who relied on `QueryRoutes` before with `num_routes` > 1, should still be able to do what they did before.

So if you used `QueryRoutes` for rebalancing with 0.6, it should still be useful to you in 0.7. It only requires you to port the k-shortest code into your application. 

In 0.6 it also didn't take into account mission control input.

**Comment by joostjager** [comment]
@C-Otto I will have a look at adding something to `queryroutes`.

**Comment by joostjager** [comment]
Can you give https://github.com/lightningnetwork/lnd/pull/3221 a try?

It first applies the ignore lists and if the node/edge is not on it, it takes a probability from mission control.

I think in the end, the route to self option is better, though it would probably also need an incoming channel restriction. We are currently only able to specify the outgoing channel.

**Comment by C-Otto** [comment]
I'd love to, but I can't figure out how to compile with Router RPC. Could you tell me?

**Comment by C-Otto** [comment]
make tags="autopilotrpc signrpc walletrpc chainrpc invoicesrpc routerrpc"

**Comment by C-Otto** [comment]
I just gave it a try. The first route had a temporary channel failure. Immediately afterwards the same route was returned by QueryRoutes, so I didn't see any change by mission control. Please tell me how to debug this, I'd be happy to help.

**Comment by C-Otto** [comment]
```
request = ln.QueryRoutesRequest(
            pub_key=pub_key,
            amt=amount,
	    use_mission_control = True,
        )

```

**Comment by joostjager** [comment]
Moving discussion into #3221

**Comment by Roasbeef** [comment]
Closed by #3221.
